### PR TITLE
[POP-76] Configure polling with feature flag

### PR DESCRIPTION
**Pull request by nitinsampath**
## What

Reading the `iso-polling-wait-intervals` feature flag value to determine how long to wait between each polling retry.

## Why

Make polling more configurable instead of waiting 1 second every poll attempt. Leads to faster response times from our clients.

## Test Plan

- ✅ | ❌ I've added unit tests for this change. 
- ✅ | ❌ The reviewer should manually test the changes in this PR.

### CODE REVIEW

File: forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt [hunk 1]
```diff
@@ -91,7 +92,8 @@ internal class CapturePaymentRepository(
         contentId: String,
         paymentRef: String
     ): ForageApiResponse<String> {
-        var attempt = 1
+        var attempt = 0
+        val pollingIntervals = LDManager.getPollingIntervals() // TODO how to get the Application and logger params
```
Comment (nitinsampath) [inline forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt:96]: how do I get `Application` and `logger` as params here

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt [hunk 2]
```diff
@@ -80,4 +99,32 @@ internal object LDManager {
...
+        } else {
+            LDConfig.Builder()
+                .mobileKey(LD_MOBILE_KEY)
+                .build()
+        }
+        val contextKind = ContextKind.of(LDContextKind.SERVICE)
+        val context = LDContext.create(contextKind, LDContexts.ANDROID_CONTEXT)
+        val client = LDClient.init(app, ldConfig, context, 0)
+
+        val jsonIntervals = client.jsonValueVariation(LDFlags.ISO_POLLING_WAIT_INTERVALS, LDValue.buildObject().put("intervals", LDValue.buildArray().add(1000).build()).build())
```
Comment (nitinsampath) [inline forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:123]: [LD](https://docs.launchdarkly.com/sdk/client-side/android/migration-2-to-3/?q=jsonvaluevariation#understanding-changes-to-the-variation-methods) introduced `LDValue` in the `jsonValueVariation` method.

We need to determine how to use `LDValue` properly here.

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt [hunk 3]
```diff
@@ -86,12 +79,25 @@ internal object LDManager {
...
+            return vaultType as String
+        }
         // default to 100% VGS usage in case LD flag retrieval fails
-        val vaultPercent = client.doubleVariation(LDFlags.VAULT_PRIMARY_TRAFFIC_PERCENTAGE_FLAG, 0.0)
+        val vaultPercent =
+            client?.doubleVariation(LDFlags.VAULT_PRIMARY_TRAFFIC_PERCENTAGE_FLAG, 0.0)
         internalLogger.i("[LaunchDarkly] Vault percent of $vaultPercent return from LD")
         val randomNum = Math.random() * 100
-        vaultType = if (randomNum < vaultPercent) {
+        vaultType = if (randomNum < vaultPercent!!) {
```
Comment (djoksimo) [inline forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:100]: doing a strict null assertion on this can cause a crash if the request to LD fails. Since we can set fallbacks, can we fallback to the default 0 value here?

**Review by djoksimo — CHANGES_REQUESTED** [review]

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt [hunk 4]
```diff
@@ -72,17 +85,54 @@ internal object LDManager {
...
     }
+
+    internal fun getPollingIntervalsFeatureFlag(): IntArray {
+        if (pollingIntervals != null) {
+            return pollingIntervals as IntArray
+        }
+
+        var pollingList = IntArray(0)
+
+        val jsonIntervals =
```
Comment (dleis612) [inline forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:122]: Could you do the same thing for this setter by including the default value here? Also, can you set the default value as a variable above for readability?

If the LDClient is null, I think this should work right?

```kotlin
val defaultVal = LDValue.buildObject().put("intervals", LDValue.buildArray().add(1000).build()).build()
val jsonIntervals = client?.jsonValueVariation(LDFlags.ISO_POLLING_WAIT_INTERVALS, defaultVal) ?: defaultVal
```

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt [hunk 5]
```diff
@@ -72,17 +85,54 @@ internal object LDManager {
...
+            client?.jsonValueVariation(LDFlags.ISO_POLLING_WAIT_INTERVALS, LDValue.buildObject().put("intervals", LDValue.buildArray().add(1000).build()).build())
```
Comment (dleis612) [inline forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:123]: What will be the behavior of the default value if it's the array `[1000]`? Are we only going to poll once after 1 second?

File: forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt [hunk 6]
```diff
@@ -90,7 +91,8 @@ internal class CapturePaymentRepository(
         contentId: String,
         paymentRef: String
     ): ForageApiResponse<String> {
-        var attempt = 1
+        var attempt = 0
+        val pollingIntervals = LDManager.getPollingIntervalsFeatureFlag() // TODO how to get the Application and logger params
```
Comment (dleis612) [inline forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt:95]: Don't believe you'll need to get the Application or logger params here anymore.

File: forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt [hunk 7]
```diff
@@ -151,8 +153,14 @@ internal class CapturePaymentRepository(
                 return ForageApiResponse.Failure(listOf(ForageError(500, "unknown_server_error", "Unknown Server Error")))
             }
 
+            var intervalTime: Long = if (pollingIntervals != null && attempt < pollingIntervals.size) {
+                pollingIntervals[attempt].toLong()
+            } else {
+                1000L
```
Comment (dleis612) [inline forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt:159]: Ah I see you default to polling for every second if there are issues with the `pollingIntervals` value.

**Review by dleis612 — CHANGES_REQUESTED** [review]

File: forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt (hunk 6 above)
Comment (nitinsampath) [reply forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt:95]: forgot to delete that comment, whoops

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt [hunk 8]
```diff
@@ -72,17 +85,54 @@ internal object LDManager {
...
             VaultType.BT_VAULT_TYPE
         } else {
             VaultType.VGS_VAULT_TYPE
         }
         internalLogger.i("[LaunchDarkly] Vault type set to $vaultType")
         return vaultType as VaultType
     }
+
+    internal fun getPollingIntervalsFeatureFlag(): IntArray {
+        if (pollingIntervals != null) {
```
[the remark this answers is not in the document]
Comment (nitinsampath) [reply forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:116]: Oh i thought this was the pattern for all feature flags in general. I can just return the latest polling intervals every time instead of storing the value in this singleton pattern

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt (hunk 4 above)
Comment (nitinsampath) [reply forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:122]: yeah that works, ill make the change

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt (hunk 5 above)
Comment (nitinsampath) [reply forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:123]: I just made the default LDValue a list of size 1 because it seems so verbose to have to do `.add(1000)` 10 times. Lemme see if there's an easier way to define a list rather than doing something like `.add(1000).add(1000).add(1000).add(1000).add(1000).add(1000).add(1000)`

File: forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt (hunk 7 above)
Comment (nitinsampath) [reply forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt:159]: yes, we'll fall back to 1 second if `pollingIntervals` was not populated correctly or if the length of the array is less than the number of max attempts.

If length of pollingIntervals is 8, but max attempts is 10, the last 2 attempts would wait 1 second each before retrying

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt (hunk 5 above)
Comment (nitinsampath) [reply forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt]: I was able to get it to work by doing 
```
        val defaultVal = LDValue.buildObject().put("intervals",
            LDValue.Convert.Integer.arrayFrom(listOf(1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000))).build()
```
This way it is more explicit what the default value is

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt [hunk 9]
```diff
@@ -33,7 +35,9 @@ internal object LDContextKind {
 internal object LDManager {
     private val LD_MOBILE_KEY = StopgapGlobalState.envConfig.ldMobileKey
     private var internalVaultType: VaultType? = null
+
     private var internalLogger: Log = Log.getSilentInstance()
+    var client: LDClient? = null
```
Comment (dleis612) [inline forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:40]: Make private por favor

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt [hunk 10]
```diff
@@ -85,4 +95,26 @@ internal object LDManager {
         internalLogger.i("[LaunchDarkly] Vault type set to $vaultType")
         return vaultType as VaultType
     }
+
+    internal fun getPollingIntervalsFeatureFlag(): IntArray {
```
Comment (dleis612) [inline forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:99]: nit: Can you remove the "FeatureFlag" portion of this function name so it's the same style as `getVaultProvider`?

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt [hunk 11]
```diff
@@ -85,4 +95,26 @@ internal object LDManager {
...
+        ).build()
+
+        val jsonIntervals = client?.jsonValueVariation(LDFlags.ISO_POLLING_WAIT_INTERVALS, defaultVal) ?: defaultVal
+
+        val intervals = jsonIntervals.get("intervals")
+
+        // Converting the LDArray into an IntArray
+        // If we cannot access LD, return int array [1000] as a default
+        for (i in 0 until intervals.size()) {
+            pollingList += intervals?.get(i)?.intValue() ?: 1000
```
Comment (dleis612) [inline forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:114]: Should be able to remove the `?: 1000` portion now right? `jsonIntervals` always has a non-null value.

File: forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt [hunk 12]
```diff
@@ -151,8 +153,14 @@ internal class CapturePaymentRepository(
                 return ForageApiResponse.Failure(listOf(ForageError(500, "unknown_server_error", "Unknown Server Error")))
             }
 
+            var intervalTime: Long = if (pollingIntervals != null && attempt < pollingIntervals.size) {
```
Comment (dleis612) [inline forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt:156]: `pollingIntervals` should be non-null now.

**Review by dleis612 — CHANGES_REQUESTED** [review]

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt [hunk 13]
```diff
@@ -51,15 +55,7 @@ internal object LDManager {
...
-    // vaultType is instantiated lazily and is a singleton. Once we set the vault type once, we don't
-    // want to overwrite it! We must take in the application as a parameter, which means that a
-    // ForagePINEditText must be rendered before any of the ForageSDKApi functions are called.
-    internal fun getVaultProvider(app: Application, logger: Log, dataSource: TestData? = null): VaultType {
-        if (vaultType != null) {
-            return vaultType as VaultType
-        }
-        internalLogger = logger
-        // Datasource is required for testing purposes!
+    internal fun setClientAndLogger(app: Application, logger: Log, dataSource: TestData? = null) {
```
Comment (djoksimo) [inline forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:58]:
```suggestion
    internal fun initialize(app: Application, logger: Log, dataSource: TestData? = null) {
```

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt [hunk 14]
```diff
@@ -85,4 +95,26 @@ internal object LDManager {
         internalLogger.i("[LaunchDarkly] Vault type set to $vaultType")
         return vaultType as VaultType
     }
+
+    internal fun getPollingIntervals(): IntArray {
```
Comment (djoksimo) [inline forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:99]: why do we return `IntArray` and then convert to long after getting the polling intervals?

Could we return a `LongArray`

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt [hunk 15]
```diff
@@ -85,4 +95,26 @@ internal object LDManager {
...
+        var pollingList = IntArray(0)
+
+        val defaultVal = LDValue.buildObject().put(
+            "intervals",
+            LDValue.Convert.Integer.arrayFrom(listOf(1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000, 1000))
```
Comment (djoksimo) [inline forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:104]:
```suggestion
            LDValue.Convert.Integer.arrayFrom(List(10) { 1000 })
```

could this work^

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt [hunk 16]
```diff
@@ -85,4 +95,26 @@ internal object LDManager {
...
+
+        val jsonIntervals = client?.jsonValueVariation(LDFlags.ISO_POLLING_WAIT_INTERVALS, defaultVal) ?: defaultVal
+
+        val intervals = jsonIntervals.get("intervals")
+
+        // Converting the LDArray into an IntArray
+        // If we cannot access LD, return int array [1000] as a default
+        for (i in 0 until intervals.size()) {
+            pollingList += intervals.get(i).intValue()
+        }
```
Comment (djoksimo) [inline forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:115]:
```suggestion
      val pollingList = IntArray(intervals.size()) { intervals.get(it).intValue() }
```

File: forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt [hunk 17]
```diff
@@ -151,8 +153,14 @@ internal class CapturePaymentRepository(
...
+            var intervalTime: Long = if (attempt < pollingIntervals.size) {
+                pollingIntervals[attempt].toLong()
+            } else {
+                1000L
+            }
+
             attempt += 1
-            delay(POLLING_INTERVAL_IN_MILLIS + getJitterAmount())
+            delay(intervalTime + getJitterAmount())
         }
```
Comment (djoksimo) [inline forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt:164]: the logic in this method seems almost the same (besides the log statements) as the logic in the balance polling. I've created a ticket to consolidate the polling logic into a reusable method/class to reduce this duplication.

https://linear.app/joinforage/issue/FX-610/consolidate-dedupe-polling-logic-in-android

**Review by djoksimo — CHANGES_REQUESTED** [review]

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt (hunk 14 above)
Comment (nitinsampath) [reply forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt:99]: Yes, then we can get rid of the `toLong()` conversion when reading the value from the list

File: forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt (hunk 17 above)
Comment (nitinsampath) [reply forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt:164]: Thanks for making this - i'll keep that change separate from this PR actually

File: forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt (hunk 16 above)
Comment (nitinsampath) [reply forage-android/src/main/java/com/joinforage/forage/android/LDManager.kt]: TIL lambda expressions exist in Kotlin!

File: forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt [hunk 18]
```diff
@@ -90,7 +91,8 @@ internal class CapturePaymentRepository(
         contentId: String,
         paymentRef: String
     ): ForageApiResponse<String> {
-        var attempt = 1
+        var attempt = 0
```
Comment (devinmorgan) [inline forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt:94]: technically, if you change `attempt` from `1` -> `0`, the if-statement `if (attempt == MAX_ATTEMPTS) { ... }` will get triggered after `MAX_ATTEMPTS + 1`. Is this OK? If we are making any gaurantees to customers that a timeout will happen in `pollingInterval * MAX_ATTEMPTS` or sooner than we may want to change the if-statement below to be `if (attempt == MAX_ATTEMPTS - 1) { ... }`.

IMO I think it would be easier to read if the if-statement remained `if (attempt == MAX_ATTEMPTS) { ... }` and we updated this code:
```kotlin
var intervalTime: Long = if (attempt < pollingIntervals.size) {
    pollingIntervals[attempt]
} else { ... }
```

to be 

```kotlin
val index = attempt - 1
var intervalTime: Long = if (index < pollingIntervals.size) {
    pollingIntervals[index]
} else { ... }
```

File: forage-android/src/main/java/com/joinforage/forage/android/network/data/CheckBalanceRepository.kt [hunk 19]
```diff
@@ -134,8 +136,14 @@ internal class CheckBalanceRepository(
                 return ForageApiResponse.Failure(listOf(ForageError(500, "unknown_server_error", "Unknown Server Error")))
             }
 
+            var intervalTime: Long = if (attempt < pollingIntervals.size) {
+                pollingIntervals[attempt]
+            } else {
```
Comment (devinmorgan) [inline forage-android/src/main/java/com/joinforage/forage/android/network/data/CheckBalanceRepository.kt:141]: Same remark as above but to this part of the code as well
```suggestion
val index = attempt - 1
var intervalTime: Long = if (index < pollingIntervals.size) {
    pollingIntervals[index]
} else {
```

**Review by devinmorgan — CHANGES_REQUESTED** [review]
Literally just the one comment (in two places). Otherwise PR looks good - appreciate all of your work on this feature, Nitin!!

File: forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt (hunk 18 above)
Comment (nitinsampath) [reply forage-android/src/main/java/com/joinforage/forage/android/network/data/CapturePaymentRepository.kt:94]: That makes sense, thanks for the good catch! I will use the `var index = attempt - 1` approach

**Review by devinmorgan — APPROVED** [review]
Nice work man! 🚢 it!

**Review by djoksimo — APPROVED** [review]
Code LGTM; nice work @nitinsampath !  

Before we release this in the next version:

Can we set the feature flag against [the staging environment and android-sdk-service](https://app.launchdarkly.com/default/staging/features/iso-polling-wait-intervals/targeting) to make sure this is working reliably end-to-end?

I'm curious about edge case test scenarios to make sure this works reliably: 

1. When the feature rule is turned off, do balance checks / QA tests succeed using the default rule?

2. When the intervals are short or depleted; will the polling interval fallback to 1000ms? [We can look at logs in datadog with timestamps](https://app.datadoghq.com/logs?query=service%3Aandroid-sdk%20&cols=host%2Cservice&index=%2A&messageDisplay=inline&refresh_mode=sliding&stream_sort=desc&viz=stream&from_ts=1696371835209&to_ts=1696372735209&live=true) to ensure this is working reliably.

```json
{ "intervals": [1] }
```

3. When the "intervals" value is invalid (with 4-eyes policy and pre-defined options, this is unlikely to happen).  Maybe something worth considering? Will launch darkly just use the default value or will balance checks cause crashes?

```json
{ "intervals": ["oops-invalid"] }

// typo
{ "interval$":  [1000] }
```
